Initial implementation of SurrealDB KVS for FoundationDB

mumoshu opened the pull request:
## What is the motivation?

I'd like to share this so that I and all potential SurrealDB users might be able to benchmark SurrealDB with TiKV and FoundationDB, which I believe the two biggest production-ready Distributed Transactional KV stores.

For more details, please read https://github.com/surrealdb/surrealdb/discussions/25.

## What does this change do?

This change adds the initial implementation of SurrealDB KVS for FoundationDB.

## What is your testing strategy?

I installed the latest stable version of FoundationDB on Ubuntu 20.04 and run all the queries explained in your [Quick start guide](https://surrealdb.com/docs/start) and saw it works.

If you want/need to follow my testing journey, here's how:

- Downloading the FDB 6.3 archive from https://apple.github.io/foundationdb/downloads.html
- Install FDB following https://apple.github.io/foundationdb/getting-started-linux.html
- `dpk -i` the archive so that FDB starts running- run `fdbcli --exec status` to ensure it's running
- Build SurrealDB and run it with the "path" to [the default FDB cluster](https://apple.github.io/foundationdb/api-c.html#database): `make build && ./target/release/surreal start --log debug --user root --pass root fdb://`
- Run queries covered in https://surrealdb.com/docs/start

## Is this related to any issues?

No, as I don't yet have an issue because this is coming from our feature request Discussion started at https://github.com/surrealdb/surrealdb/discussions/25 and the current rule is to use Discussions instead of Issues for a feature request.

I'm happy to submit an Issue if you need something that can't get covered by https://github.com/surrealdb/surrealdb/discussions/25 though.

## Have you read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/master/CONTRIBUTING.md)?

[x] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/master/CONTRIBUTING.md)

I did read it, and https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md#scalability-and-performance is why I tried to replicate the look&feel of the SurrealDB KVS for TiKV in this new FDB KVS so that I don't do anything that is obviously bad compared to the TiKV impl.

Also, https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md#external-dependencies is why I tried not to introduce anything other than the [foundationdb-rs](https://github.com/foundationdb-rs/foundationdb-rs) into the direct dependencies of the project. Note though that I did introduce transitive dependencies of `bindgen` and `env_logger` along with the `foundationdb-rs`.

I'm unsure what to do for https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md#submitting-a-pull-request in this specific case- if you think it's better to make the branch name `discussion-25-fdb-kvs` I'd happily do so!

tobiemh reviewed (changes requested):
Hi @mumoshu thanks for this! Absolutely awesome! I've made a couple of minor comments on the code itself.

In addition to those changes, can you run `cargo fmt` on the repository, just to format the code consistently? I'll add this to the contributing docs so apologies for that.

The one issue that remains, is that this doesn't build on `macOS arm64`. As a result it would break the continuous integration build and release. I'm thinking for the moment that we remove the `kv-fdb` feature from the default features (while I make some changes so that only the Linux build picks up the FoundationDB implementation).

What are your thoughts?

mumoshu commented:
@tobiemh Hey! Thanks a lot for reviewing 🙏 

> run cargo fmt on the repository

Done.

> remove the kv-fdb feature from the default features

Absolutely fine to do so!
Thanks for pointing out that it doesn't build for M1 mac.

> tokio::sync dep

Ah, I forgot that I added it! Sorry about that.
I've replaced it with `futures::lock::Mutex` and removed the new tokio dependency out of the lib module.

Here's the commit that I made to address those: [bdefe2b](https://github.com/surrealdb/surrealdb/pull/37/commits/bdefe2b18996d3e7e8f8088cc1b5b42ea6768304)

Would you mind taking a look once again?

mdecimus commented:
Hi @tobiemh and @mumoshu 

I was looking at the FoundationDB datastore implementation in the SurrealDB `main` branch and I noticed that no chunking is being done when values exceeds the FDB [100,000 bytes limit](https://apple.github.io/foundationdb/known-limitations.html). 

For example, in `TermDocs::set_doc` you are serializing a `RoaringTreemap` and then storing it in the database. If the serialized value exceeds 100kb (which is [quite likeley](https://github.com/RoaringBitmap/RoaringBitmap/blob/master/README.md#faq) when serializing hundreds of thousands of values), the entire transaction will fail.

In addition to this, the FDB code does not seem to be checking whether the entire transaction exceeds the [10MB size limit](https://github.com/apple/foundationdb/wiki/Transaction-size-limit).

Am I missing something or is the FDB connectors not taking into consideration these limitations?

mumoshu commented:
@mdecimus Hi! Thank you very much for sharing and asking!

Absolutely; You are correct that we don't currently and automatically chunk KV pairs depending on the KVS used.

I considered, and still consider, that it is not the responsibility of the current `Datastore` layer of SurrealDB, which abstracts away various KVS.
That's because chunking needs different key prefixes or some ad-hoc hack; Otherwise, any range scan returns KV pair chunks rather than the asked KV pairs.

// Note: FDB is well known for having limitations. But even TiKV has its limit. See https://github.com/tikv/tikv/issues/7272#issuecomment-604841372 for example. So, we need to be aware of chunking regardless of whether you use FDB.

In the future, the consumers of `Datastore` need to be responsible for chunking. They basically assume any KVS that SurrealDB supports today can handle infinitely-sized values.

Implementation-wise, maybe we could introduce an intermediate layer between, e.g., SurrealDB's indexing subsystem (who cares about `TermDocs::set_doc` and `RoaringBitmap` you mentioned) and Datastore, something like `ChunkAwareDatastore`(better name wanted 😂) which configurable chunk size.

> In addition to this, the FDB code does not seem to be checking whether the entire transaction exceeds the [10MB size limit](https://github.com/apple/foundationdb/wiki/Transaction-size-limit).

Good point! This is currently by design. An important note is that although FDB (and TiKV) has some preconfigured transaction size limit, it doesn't mean other KVS can support infinitely large transactions.

So the best we could do is to have some continuous large-scale testing in SurrealDB and catch cases like this one(not having chunking in a needed place) and fix it.

This way, we could say, "SurrealDB doesn't impose any transaction and KV size-related restriction on its own. The limit is that of the underlying KVS you use for your SurrealDB deployment.`.

@mdecimus Does the above make sense to you?

I am looking forward to hearing from you!
In the meantime, Thank you again for sharing your thoughts and asking!
I'm happy that we can talk about anything related to transactions, as that is one of the things I deeply care about in SurrealDB 💟
